File: common/src/main/java/me/kvalbrus/multibans/common/punishment/punishments/MultiTemporaryBanIp.java
package me.kvalbrus.multibans.common.punishment.punishments;

import java.util.List;
import me.kvalbrus.multibans.api.OnlinePlayer;
import me.kvalbrus.multibans.api.punishment.creator.OnlinePunishmentCreator;
import me.kvalbrus.multibans.api.punishment.creator.PunishmentCreator;
import me.kvalbrus.multibans.api.punishment.target.PunishmentTarget;
import me.kvalbrus.multibans.api.punishment.PunishmentType;
import me.kvalbrus.multibans.api.punishment.punishments.TemporaryBanIp;
import me.kvalbrus.multibans.common.managers.PluginManager;
import me.kvalbrus.multibans.common.permissions.Permission;
import me.kvalbrus.multibans.common.punishment.MultiTemporaryPunishment;
import me.kvalbrus.multibans.common.utils.Message;
import me.kvalbrus.multibans.common.utils.ReplacedString;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MultiTemporaryBanIp extends MultiTemporaryPunishment implements TemporaryBanIp {

    public MultiTemporaryBanIp(@NotNull PluginManager pluginManager,
                               @NotNull String id,
                               @NotNull PunishmentTarget target,
                               @NotNull PunishmentCreator creator,
                               long createdDate,
                               long startedDate,
                               long duration,
                               @Nullable String reason,
                               @Nullable String comment,
                               @Nullable PunishmentCreator cancellationCreator,
                               long cancellationDate,
                               @Nullable String cancellationReason,
                               @NotNull List<String> servers,
                               boolean cancelled) {
        super(pluginManager, PunishmentType.TEMP_BAN_IP, id, target, creator, createdDate,
            startedDate, duration, reason, comment, cancellationCreator, cancellationDate,
            cancellationReason, servers, cancelled);
    }

    @Override
    public synchronized void activate() {
        super.activate();

        // Sends a message to all players with permission
        ReplacedString listenMessage = new ReplacedString(Message.TEMPBANIP_ACTIVATE_LISTEN.getMessage())
            .replacePunishment(this);

        for (OnlinePlayer p : this.getPluginManager().getOnlinePlayers()) {
            if (p.hasPermission(Permission.PUNISHMENT_TEMPBANIP_LISTEN.getName())) {
                p.sendMessage(listenMessage.string());
            }
        }

        // Sends a message to console
        this.getPluginManager().getConsole().sendMessage(listenMessage.string());


        // Sends a message to the creator
        if (this.getCreator() instanceof OnlinePunishmentCreator creator) {
            ReplacedString creatorMessage = new ReplacedString(Message.TEMPBANIP_ACTIVATE_EXECUTOR.getMessage())
                .replacePunishment(this);
            creator.sendMessage(creatorMessage.string());
        }
    }

    @Override
    public synchronized void deactivate(@Nullable PunishmentCreator cancellationCreator,
                                        long cancellationDate,
                                        @Nullable String cancellationReason) {
        super.deactivate(cancellationCreator, cancellationDate, cancellationReason);

        // Sends a message to all players with permission
        ReplacedString listenMessage = new ReplacedString(Message.TEMPBANIP_DEACTIVATE_LISTEN.getMessage())
            .replacePunishment(this);

        for (OnlinePlayer p : this.getPluginManager().getOnlinePlayers()) {
            if (p.hasPermission(Permission.PUNISHMENT_TEMPBANIP_LISTEN.getName())) {
                p.sendMessage(listenMessage.string());
            }
        }

        // Sends a message to console
        this.getPluginManager().getConsole().sendMessage(listenMessage.string());

        // Sends a message to the creator
        if (this.getCreator() instanceof OnlinePunishmentCreator creator) {
            ReplacedString creatorMessage = new ReplacedString(Message.TEMPBANIP_DEACTIVATE_EXECUTOR.getMessage())
                .replacePunishment(this);
            creator.sendMessage(creatorMessage.string());
        }
    }

    @Override
    public synchronized void delete() {
        super.delete();

        // Sends a message to all players with permission
        ReplacedString listenMessage = new ReplacedString(Message.TEMPBANIP_DELETE_LISTEN.getMessage())
            .replacePunishment(this);

        for (OnlinePlayer p : this.getPluginManager().getOnlinePlayers()) {
            if (p.hasPermission(Permission.PUNISHMENT_TEMPBANIP_LISTEN.getName())) {
                p.sendMessage(listenMessage.string());
            }
        }

        // Sends a message to console
        this.getPluginManager().getConsole().sendMessage(listenMessage.string());

        // Sends a message to the creator
        if (this.getCreator() instanceof OnlinePunishmentCreator creator) {
            ReplacedString creatorMessage = new ReplacedString(Message.TEMPBANIP_DELETE_EXECUTOR.getMessage())
                .replacePunishment(this);
            creator.sendMessage(creatorMessage.string());
        }
    }
}

File: common/src/main/kotlin/me/kvalbrus/multibans/common/punishment/punishments/MultiTemporaryBanIp.kt
package me.kvalbrus.multibans.common.punishment.punishments

import me.kvalbrus.multibans.api.punishment.action.Action
import me.kvalbrus.multibans.api.punishment.action.CreationAction
import me.kvalbrus.multibans.api.punishment.punishments.PunishmentType
import me.kvalbrus.multibans.api.punishment.punishments.TemporaryBanIp
import me.kvalbrus.multibans.common.managers.PluginManager
import me.kvalbrus.multibans.common.permissions.Permission
import me.kvalbrus.multibans.common.utils.Message.*
import me.kvalbrus.multibans.common.punishment.MultiTemporaryPunishment
import me.kvalbrus.multibans.common.utils.Message

final class MultiTemporaryBanIp : MultiTemporaryPunishment, TemporaryBanIp {

    constructor(
        pluginManager: PluginManager,
        id: String,
        creationAction: CreationAction,
        activations: MutableList<Action> = mutableListOf(),
        deactivations: MutableList<Action> = mutableListOf(),
        startedDate: Long,
        duration: Long,
        comment: String,
        servers: List<String>,
        cancelled: Boolean) : super(pluginManager, PunishmentType.TEMP_BAN_IP, id, creationAction,
        activations, deactivations, startedDate, duration, comment, servers, cancelled)

    override val activateMessageForListener: Message = TEMPBANIP_ACTIVATE_LISTEN
    override val activateMessageForExecutor: Message = TEMPBANIP_ACTIVATE_EXECUTOR
    override val activateMessageForTarget: Message = TEMPBANIP_ACTIVATE_TARGET
    override val deactivateMessageForListener: Message = TEMPBANIP_DEACTIVATE_LISTEN
    override val deactivateMessageForExecutor: Message = TEMPBANIP_DEACTIVATE_EXECUTOR
    override val deactivateMessageForTarget: Message = EMPTY
    override val createMessageForListener: Message = TEMPBANIP_ACTIVATE_LISTEN
    override val createMessageForExecutor: Message = TEMPBANIP_ACTIVATE_EXECUTOR
    override val createMessageForTarget: Message = TEMPBAN_ACTIVATE_TARGET
    override val deleteMessageForListener: Message = TEMPBANIP_DELETE_LISTEN
    override val deleteMessageForExecutor: Message = TEMPBANIP_DELETE_EXECUTOR
    override val deleteMessageForTarget: Message = TEMPBANIP_DELETE_TARGET
    override val reasonChangeMessageForExecutor: Message = TEMPBANIP_REASON_CREATE_CHANGE_EXECUTOR
    override val reasonChangeMessageForListener: Message = TEMPBANIP_REASON_CREATE_CHANGE_LISTEN
    override val commentChangeMessageForExecutor: Message = TEMPBANIP_COMMENT_CHANGE_EXECUTOR
    override val commentChangeMessageForListener: Message = TEMPBANIP_COMMENT_CHANGE_LISTEN
    override val permissionForListener: Permission = Permission.PUNISHMENT_TEMPBANIP_LISTEN
}